Introduce Cookbook (#424)
hyperledger/iroha-2-docs: src/cookbook/work-with-non-mintable-assets.md

---
title: "Work with Non-Mintable Assets | Cookbook"
head:
  - - meta
    - name: description
      content: "Learn how to work with non-mintable assets in Iroha."
  - - meta
    - name: keywords
      content: "non-mintable assets, Iroha assets"
---

# How to Work with Non-Mintable Assets

```rust
fn register_non_mintable_asset(
    iroha: &Client,
) {
    let magical_keys = AssetDefinitionId::from_str("magical_key#wonderland").unwrap();
    // register keys as an asset definition
    let register_keys_as_a_concept = Register::asset_definition(
        AssetDefinition::new(
            magical_keys.clone(),
            AssetValueType::Numeric(NumericSpec::integer()),
        ).mintable_once()
    );
    let alice = "alice@wonderland".parse().unwrap();
    // Alice owns ten keys and cannot mint more
    let initial_keys_of_alice = Asset::new(
        AssetId::new(roses, alice), 
        10_u32
    );
    let register_keys_of_alice = Register::asset(initial_keys_of_alice);

>     iroha.submit_all([
>         InstructionBox::from(register_keys_as_a_concept),
>         InstructionBox::from(register_keys_of_alice)
>     ]).unwrap();

Reviewer: ```suggestion
    iroha.submit_all([register_keys_as_a_concept, register_keys_of_alice]).unwrap();
```

this works because both instructions are of the same type

Reviewer 2: The generics are different, one is `Register<AssetDefinition>`, the other is `Register<Asset>`. However, we can change `InstructionBox` to `RegisterBox`

Reviewer: ok, then:
```suggestion
    let instructions: [InstructionBox; _] = [
        register_keys_as_a_concept.into(),
        register_keys_of_alice.into()
    ];
    iroha.submit_all(instructions).unwrap();
```

}
```

```rust
fn mint_non_mintable_asset(
    iroha: &Client,
) {
    // Alice owns zero keys and can mint once
    let keys_of_alice = "magical_key##alice@wonderland".parse().unwrap();
    let zero_keys_of_alice = Asset::new(
        keys_of_alice.clone(), 
        0_u32
    );
    let register_keys_of_alice = Register::asset(zero_keys_of_alice);
    let mint_keys_for_alice = Mint::asset_numeric(10_u32, keys_of_alice);

>     iroha.submit_all([
>         InstructionBox::from(register_keys_as_a_concept),
>         InstructionBox::from(register_keys_of_alice),
>         InstructionBox::from(mint_keys_for_alice),
>     ]).unwrap();

Reviewer: ```suggestion
    let instructions: [InstructionBox; _] = [
        register_keys_as_a_concept.into(),
        register_keys_of_alice.into(),
        mint_keys_for_alice.into(),
    ];
    iroha.submit_all(instructions).unwrap();
```

# How to Work with Non-Mintable Assets

```rust
fn register_non_mintable_asset(
    iroha: &Client,
) {
    let magical_keys = AssetDefinitionId::from_str("magical_key#wonderland").unwrap();
    // register keys as an asset definition
    let register_keys_as_a_concept = Register::asset_definition(
        AssetDefinition::new(
            magical_keys.clone(),
            AssetValueType::Numeric(NumericSpec::integer()),

>         ).mintable_once()

Author: This snippet registers an asset with `mintable_once`, which is different from a non-mintable asset.

Reviewer 2: I took the example from the integration tests: https://github.com/hyperledger/iroha/blob/v2.0.0-pre-rc.21.1/client/tests/integration/non_mintable.rs#L12

    );
    let alice = "alice@wonderland".parse().unwrap();
    // Alice owns ten keys and cannot mint more
    let initial_keys_of_alice = Asset::new(
        AssetId::new(roses, alice), 
        10_u32
    );
    let register_keys_of_alice = Register::asset(initial_keys_of_alice);
    iroha.submit_all([
        InstructionBox::from(register_keys_as_a_concept),
        InstructionBox::from(register_keys_of_alice)
    ]).unwrap();
}
```

```rust

> fn mint_non_mintable_asset(

Author: If it's non-mintable, how it could be minted 🤔

Reviewer 2: Good question, I think the mintability naming is unfortunate, and we should rethink it. But that's an issue for the main repo

Reviewer 2: Same as https://github.com/hyperledger/iroha-2-docs/pull/424#discussion_r1607990016
